hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataResolver.java

```diff
@@ -56,59 +60,113 @@ public Mapping resolve(String iMapName) {
             return null;
         }
 
-        boolean hd = container.getMapConfig().getInMemoryFormat() == InMemoryFormat.NATIVE;
-        Metadata metadata = hd ? resolveFromHd(container) : resolveFromHeap(iMapName, context);
+        boolean hd = container.getMapConfig().getInMemoryFormat() == InMemoryFormat.NATIVE
+                && !container.getMapConfig().getTieredStoreConfig().isEnabled();
+
+        Metadata metadata = hd
+                ? resolveFromContentsHd(iMapName, context)
+                : resolveFromContents(iMapName, context);
         return metadata == null
                 ? null
                 : new Mapping(iMapName, iMapName, null, TYPE_NAME, null, metadata.fields(), metadata.options());
     }
 
     @Nullable
-    @SuppressWarnings("rawtypes")
-    private Metadata resolveFromHd(MapContainer container) {
-        if (container.getIndexes() == null) {
-            return null;
+    private Metadata resolveFromContents(String name, MapServiceContext context) {
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            RecordStore<?> recordStore = partitionContainer.getExistingRecordStore(name);
+            Metadata resolved = resolveFromContentsRecordStore(nodeEngine, recordStore);
+            if (resolved != null) {
+                return resolved;
+            }
         }
+        return null;
+    }
 
-        InternalIndex[] indexes = container.getIndexes().getIndexes();
-        if (indexes == null || indexes.length == 0) {
-            return null;
+    private Metadata resolveFromContentsHd(String name, MapServiceContext context) {
+        // Iterate only over local partitions.
+        // MC will invoke this operation on each member until it finds some data.
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
```

Comment:
scanning owned or backup partitions should be sufficient, otherwise many unneeded operation call can happen.